### DISCUSSION TITLE
Ultima 6 Mod is Out!

### DISCUSSION THREAD

**Discussion by KnightCaptainU7** [exult/exult#345] [Show and tell]
Looks like the Ultima 6 mod was made available earlier this month.
https://www.ultima6.realmofultima.com/

**Comment by xboxown** [comment]
I am having hard time getting it run. I keep getting back to ultima 7 black. Even if I selected ultima 6 mode from the list it goes right to black gate. Any advise how I can fix this issue?

**Reply by xboxown** [reply]
I got it working. I did.

**Reply by KnightCaptainU7** [reply]
What was the issue and fix?

**Reply by xboxown** [reply]
Nothing, extract it like this. 

![image](https://github.com/exult/exult/assets/14060186/9fb6c3c3-053d-4b59-967a-5fa19993c198)

**Reply by xboxown** [reply]
Speaking of ultima 7...dupre died and I need to find his body. Excuse me while I hunt for his body.

I LOVE ULTIMA 7 ENGINE SO MUCH!!! BEST GAME ENGINE ON PLANET EARTH!

**Reply by KnightCaptainU7** [reply]
Yea, I don't know why it didn't come out as a Mod vs. a Patch.
Exult is faithful to the originals, and losing the corpse can happen there.

**Reply by Tycho-X** [reply]
Hi,

thank you for this great mod. Unluckily i did run into a bug just when exiting LBs castle. The group is running into "Finn" and the conversation restarts endlessy after using "bye". When i try to run away exult does a coredump. This are the last lines i can see in my terminal.

[code]
Npc is not in WAIT schedule, set to TALK
Stack underflow
Iolo counted 67 gold coins in the party.
Stack underflow
Stack underflow
Npc is not in WAIT schedule, set to TALK
Stack underflow
Stack underflow
Stack underflow
chaosSword called
Stack underflow
Stack underflow
Npc is not in WAIT schedule, set to TALK
Stack underflow
Stack underflow
Stack underflow
Npc is not in WAIT schedule, set to TALK
Stack underflow
Stack underflow
(x,y) rel. to gump is (119, 69)
Stack underflow
Finding path to schedule for Finn
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Stack underflow
Shape::store_frame:  framenum >= frames.size() (5 >= 5)
Shape::store_frame:  framenum >= frames.size() (7 >= 5)
Shape::store_frame:  framenum >= frames.size() (6 >= 5)
Segmentation fault (core dumped)
[/code]

Anything i can do to help debugging this?

**Comment by xboxown** [comment]
Well, the engine is so sweet, I have decided to make real editor for Ultima 7. I am building the engine from scratch, doing some improvement to the engine and when my engine is 100% complete, I will work from scratch a complete GUI editor that will be as friendly as RPG Maker. With this editor it will be all click and play even script just like RPG Maker and people can make their games from scratch...not needing ultima 7 data file. You make everything from scratch, with your own graphics, music, sound, story, plot, blah blah.

My intention is making an entire new game from scratch using the same style Ultima 7 with the improvement...such as you need to be close to an object to pick something up or talk to NPC unlike Ultima 7 where you can do it from far as if you have permanent telekinesis. 

In the meantime, I am enjoying ultima 6 using ultima 7 it is like a third ultima 7 game released. I love concept of losing items, or having to find where you last dropped your item, or searching for your body etc. By the way, in Ultima 6, I am thinking I may have asked Dupre to leave party. If that is the case where would I go to ask him to join me again (in case, I asked him to leave)?

**Reply by KnightCaptainU7** [reply]
You may have to use the F2 cheat menu to find his location. I would check the local tavern first.

**Reply by xboxown** [reply]
If I press f2 where do i go from here?
assume he is dead, will it show me his body location?

**Comment by DominusExult** [comment]
Could you please keep on topic and not diverge a topic in several non related topics?

**Comment by KnightCaptainU7** [comment]
I've reached out to AgentOrangeMan to give a little feedback. Hopefully we'll see the codebase at some point.

**Comment by exusninja** [comment]
anyone have an issue with crashing game after speaking to NPCs CHuckles and SHerry

**Reply by DominusExult** [reply]
There was a bug in Exult that Marzo fixed earlier today.
Tomorrow's snapshot should have it!

**Comment by Transmodiar** [comment]
Are there any instructions on how to install this on MacOS? Mods go in Library > Application Support > Exult > blackgate > mods, right? I've got the Keyring mod in there and it shows up no problem, but moving the Ultima VI folder over does nothing.

I worked on the Ultima VI Project and would very much like to see how Exult was used to recreate the game!

**Comment by Dragon-Baroque** [comment]
> I've got the Keyring mod in there and it shows up no problem, but moving the Ultima VI folder over does nothing.

Please take care that you also have to copy the `Ultima6v1.1.cfg` one folder up : from `mods/Ultima6v1.1/` into `mods/`.
So that you get, for example, in the `mods/` folder :
 - `Keyring/`
 - `Keyring.cfg`
 - `Ultima6v1.1/`
 - `Ultima6v1.1.cfg`
 - ...
 
**NB** Also please notice that Ultima VII the Black Gate and all its mods should go into `forgeofvirtue/` not into `blackgate/` unless you own the original game without the Forge of Virtue on diskettes, and same with Serpent Isle into `silverseed/` not into `serpentisle/`.

**Comment by i30817** [comment]
This mod freezes exult aleatory once you get 8 (or maybe less) party members. It's very obvious once you reach the limit, only save game spamming and resetting will make you progress by resetting the last save and advance a bit further, then save, repeat. 

I'm almost certain my last save had more than the core avatar+3 party + 1 temp party member of exult as a result of this save spamming when I was experimenting with this bug. Unfortunately I can't debug ATM, but I'm in a old android 10 to be clear, arm64, using the release not a debug snapshot, although exult released a minor version 5 days ago anyway, so it's recent. 

I suspect the pathfinding of the follow ai gets into a loop (not the double click avatar pathfinding, the 'party members follow the avatar but displace each other' where the pathfinding is waiting to displace one that is waiting to displace another that is waiting to displace the first (or a bigger chain, dining philosophers problem). Maybe it can only happen when the Avatar is affected (because they are the prime mover of the chain and if they could keep moving even if the others were stuck, they would, or maybe the movement loop never quite gets to the avatar again for some reason, since the engine is single threaded).

I suspect this because it's one obvious thing affected by more party members and because I have the suspicion it happens more often in corridors and doors and near walls and changes of direction when the party bunches up from its usual triangle formation.

**Comment by i30817** [comment]
Also you probably already know but for Linux, and because exult doesn't have case insensitive loading of data files yet, I had to lower case all files because there are some rogue files in the mod even after release 1.2 where in Linux it manifests as broken buildings walls and ceilings. Easy to fix if you realize what's the problem but easier to fix in the source of the mod for now imo.

**Comment by i30817** [comment]
I don't know who here is involved with the mod so I can't notify them and my bug report about the first problem in exult was closed (even if I disagree with that this is not a exult bug) but here is my report for what's it worth. A shout into the void.

**Reply by wench** [reply]
No one here is involved with the mod. If you can replicate the problem in unmoded BG, make an issue for it

**Reply by i30817** [reply]
You can't. However, it _could_ follow from code in exult if it is what I think, because if the algorithm to prevent stuck 'philosophers' was specific to at most 4 party members not including the avatar because in both games those are the max at any one time. 

If this is the case and the u6 mod isn't using mod specific script pathfinding which I find unlikely, maybe using exult cheats to add 8 party members into forge of virtue (the base game for the u6 mod) in exult would exhibit the bug. If that's possible that is.

**Reply by KnightCaptainU7** [reply]
Black Gate / Forge of Virtue allows for up to 8 party members out of the box, as did Ultima 6. The limit of 5 was set for Serpent Isle / Silver Seed, but this limit was set by Usecode.

Have you tried the link at the top of the thread to report the problem to them?

**Reply by i30817** [reply]
Can't register in disquis for some reason.

**Reply by GrantWChapman** [reply]
I've sent the creator a message that you had a bug report for them on the Ultima Dragons Discord.

**Comment by agentorangeguy** [comment]
I'm not sure why it's crashing but nothing in the mod affects party size.  The only thing I can think of is that maybe if you had a party member leave it might still show them technically in the party, but other than that I'm not sure.  Did you make any of them leave the party?

**Reply by i30817** [reply]
Well, I recruited sherry, which technically may mean that the party temporarily left (for when sherry goes into the hole to get the plot item)... 

But I'm not really sure this is actually needed (I think not, just adding enough party, iirc). And note it's not a crash, it's a freeze of exult. I can minimize exult in android and pull it back again, the difference is that it stops rendering the frozen image (it's not a black screen because the joystick and esc for Android exult interface still appear and even move\press, I suspect because they're being rendered by java\another thread technically). That's why I suggested a lock\infinite loop, problem. 


After reproducing where it is (and it's REALLY easy with a large party and switching directions of movement near walls), if you can compile figuring out what is looping should be quite possible by using a debugger and breaking by shortcut and looping until you get to the original place you breaked at (you can even use RR reversible debugger to check out the initial loop conditions that screwed up).

I can't because I'm 1. out of habit 2. don't have a Linux machine\rooted android.

**Reply by i30817** [reply]
> I'm not sure why it's crashing but nothing in the mod affects party size.  The only thing I can think of is that maybe if you had a party member leave it might still show them technically in the party, but other than that I'm not sure.  Did you make any of them leave the party? 

My current save has the avatar, shamino, iolo, dupre, sherry, gwenno and Julia, in order. I don't know how to post a exult android save game.

**Reply by agentorangeguy** [reply]
Does this happen on u7 black gate? or just the mod? I still don't know what it could be

**Reply by i30817** [reply]
Well, I haven't tried it yet, but may do in the near future (I'm having some home improvement projects), but I suspect not because I'd expect a lot of people tried Android exult over the years with blackgate\forgeofvirtue.

But who knows, it may be a new bug.

**Reply by i30817** [reply]
> Does this happen on u7 black gate? or just the mod? I still don't know what it could be

I now tried (in blackgate forge of virtue) after using the cheat menu to teleport to party members (and with the annoyance that some but not all of the party members refuse to join if a party of 7 so you have to make someone leave temporarily to get 8), and I can't make it freeze like the mod (with just 7). 

Well I only tried running around for a minute but it's still didn't happen, unlike the in the mod where it's pretty predictable. I wish I could send the saves. 

I'm also pretty annoyed that I never realized that the party limit is larger than I ever bothered to look at in blackgate (to be fair, I mostly played serpent isle over the years, the system of ethics is fascinating in a way that the virtues simply weren't to me).

**Reply by agentorangeguy** [reply]
Sherry doesn't actually leave the party during the cutscene.  Where exactly are you having this issue and I will try to recreate it

**Reply by i30817** [reply]
it happens anywhere when you have 7 party members. I also restarted before trying again and it also happened.

**Comment by i30817** [comment]
btw there is another bug where I'm unable to talk to sherry unless I do it just after having loaded a save near her. Even using the click to show menu, selecting sherry and then talk buttons. 

You really only need to save then reload near her. But it's kind of weird.

**Reply by agentorangeguy** [reply]
That one is fixed in the soon to be released latest version.  I will probably release the usecode first since Sherry is very buggy but there was just a line of code that was making it wonky and has been fixed.

**Reply by i30817** [reply]
I've been wondering why is the book to solve the wizard of oz quest for lord British on his room table? It does make Lord Brit look more foolish, which is sort of a good thing, but I think it's a bit too much. I've read that originally it was in the lyceum (which kind of introduces the problem of how it got there because only some brittainians are from earth, was that why you changed it?

**Comment by wench** [comment]
Been playing this lately and the mod caused Exult to crash when trying to hatch eggs in Sutak's dungeon. I can't tell exactly why it crashed but it looks like for some reason a non egg object got added to one of Exult's chunk egg_objects vector and things got screwed up by a bad vtable. I can certainly fix the crash but I would rather figure out how to make it be repeatable so I can figure out why this could even happen in the first place and why exult would try to treat a non egg as an egg. The bad object in the crash dump is just strange, its chunk pointer is null. Maybe something related to multimap not working properly as this is really the first mod that really heavily uses multimap. Thinking more about it, I think it is an object that was destructed as only the vtable looks wrong for an egg and the smart pointer has a ref count of zero  so my crash was a use after free. Still want to know why.

**Comment by Allanon41** [comment]
i can't seem to use the new mod even though i have everything in the right place. im on iOS and i have the mods folder in the forgeofvirtue folder and the mod ultima6v1.2.cfg next to the mod folder (ultima6v1.2). i see the mod displayed after tapping mods on bootup, i tap the Ultima v 1.6 name but it goes to ultima 7. the mod folder was originally named patch after unzipping, but i renamed to ultima6v1.2. could that be the problem? or is it more appropriate in a patch configuration? does cfg file itself need to point to the mod or the disply name reference need to be the same as mod folder name?

**Reply by DominusExult** [reply]
You likely messed up the paths:
In the game/forgeofvirtue folder should be three subfolders: STATIC, patch, mods.
The u6 mod needs to go inside the mods folder. 
1. remove your previous attempt.

2. Grab the mod from our dl site https://exult.info/snapshots/Ultima6.zip

3. Unzip that into the mods folder without renaming anything.
In the mods folder you should now have the Ultima6v1.2.cfg and a folder named Ultima6v1.2

If you download the file on your iOS device you can do this via the files app (move the zip into the mods folder and unzip it in there - it's less error prone than moving the files from another folder - the files app can choke on many small files operations 🤷‍♂️)

**Reply by Allanon41** [reply]
i did that and it gave an ultima6 folder (iOS always defaults unzips into a folder of the name of the zip). but i moved those out and into mods folder and it now works. i must misnamed something, though at first it did show ultima vii intro before going into ultima 6. and trying to view ultima 6 intro from u6 main menu still gives me u7 intro.
